### PR TITLE
Adding prefix operators.

### LINKED ISSUE #244: Precedences for prefix operators
It would be very convenient for Lambdapi to support custom precedences for prefix operators, similar to the existing support for infix operators. Note that it is possible to define prefix operators, as follows:

~~~
definition not : eta (o > o) ≔ λ p, p --> false
set declared "~"
definition ~ : eta (o > o) ≔ not
~~~

However, the operator `~` then always binds just like any other function symbol, such as `not`. This really shows in the following example:

~~~
// these expressions typecheck
type λ p, ~ p
type λ p, ~ (~ p)
type λ p x, ~ (p x)

// these expressions do not typecheck, but I would like them to!
type λ p, ~ ~ p
type λ p x, ~ p x
~~~

Such expressions regularly appear in Coq and Isabelle. I ported some Coq proofs to Lambdapi, and forgetting to put parentheses around negated expressions was by far my most frequent error (this alone probably cost me hours).

I would like to have something like:

~~~
definition not : eta (o > o) ≔ λ p, p --> false
set prefix 40 "~" ≔ not
~~~

Similar syntax in Isabelle:

~~~
definition not :: "o ⇒ o"  ("¬ _" [40] 40)
  where "not ≡ λA. A ⟶ False"
~~~

### PR DESCRIPTION
**Pull request by rlepigre**
Fixes #244.

### CODE REVIEW

File: src/handle.ml [hunk 1]
```diff
@@ -267,6 +267,12 @@ let handle_cmd : sig_state -> p_command -> sig_state * proof_data option =
             Sign.add_builtin ss.signature s sym;
             out 3 "(conf) set builtin [%s]\n" s;
             {ss with builtins = StrMap.add s sym ss.builtins}
+        | P_config_unop(unop)     ->
+            let (s, _, qid) = unop in
+            (* Define the binary operator [s]. *)
```
Comment (fblanqui) [inline src/handle.ml]: binary -> unary

File: src/parser.ml [hunk 2]
```diff
@@ -86,6 +86,12 @@ module Prefix :
       Earley.black_box fn Charset.full false "<tree>"
   end
 
+(** Currently defined binary operators. *)
```
Comment (fblanqui) [inline src/parser.ml]: binary -> unary

File: src/parser.ml [hunk 3]
```diff
@@ -100,21 +106,23 @@ let declared_id = Prefix.grammar declared_ids
...
   [ "("; ")"; "."; "λ"; "∀"; "&"; "["; "]"; "{"; "}"; "?"; "=" ; ":"; "→"
   ; "@"; ","; ";"; "\""; "\'"; "≔"; "//"; " "; "\r"; "\n"; "\t"; "\b" ]
   @ List.init 10 string_of_int
 
 (** [get_binops loc p] loads the binary operators associated to module [p] and
     report possible errors at location [loc].  This operation requires the [p]
     to be loaded (i.e., compiled). The declared identifiers are also retrieved
-    at the same time. *)
+    at the same time, as well as the unary operators. *)
 let get_binops : Pos.pos -> p_module_path -> unit = fun loc p ->
```
Comment (fblanqui) [inline src/parser.ml]: rename into get_ops

File: src/parser.ml [hunk 4]
```diff
@@ -100,21 +106,23 @@ let declared_id = Prefix.grammar declared_ids
 
 (** The following should not appear as substrings of binary operators, as they
     would introduce ambiguity in the parsing. *)
-let forbiden_in_binops =
+let forbiden_in_ops =
   [ "("; ")"; "."; "λ"; "∀"; "&"; "["; "]"; "{"; "}"; "?"; "=" ; ":"; "→"
   ; "@"; ","; ";"; "\""; "\'"; "≔"; "//"; " "; "\r"; "\n"; "\t"; "\b" ]
```
[the remark this answers is not in the document]
Comment (fblanqui) [reply src/parser.ml]: What about "\'" ?

[the remark this answers is not in the document]
Comment (rlepigre) [reply src/parser.ml]: I'm anticipating for a potential character type.

**Comment by 01mf02** [comment]
Thank you, this works perfectly!

**Pull request merged**